packages/flutter/test/widgets/app_lifecycle_listener_test.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  Future<void> setAppLifeCycleState(AppLifecycleState state) async {
+    final ByteData? message = const StringCodec().encodeMessage(state.toString());
+    await ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage('flutter/lifecycle', message, (_) {});
+  }
+
+  Future<void> sendAppExitRequest() async {
+    final ByteData message = const JSONMethodCodec().encodeMethodCall(const MethodCall('System.requestAppExit'));
+    await ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage('flutter/platform', message, (_) {});
+  }
+
+  setUp(() async {
+    WidgetsFlutterBinding.ensureInitialized();
+    WidgetsBinding.instance
+      ..resetEpoch()
+      ..platformDispatcher.onBeginFrame = null
+      ..platformDispatcher.onDrawFrame = null;
+    final TestWidgetsFlutterBinding binding = TestWidgetsFlutterBinding.instance;
+    binding.platformDispatcher.resetInitialLifecycleState();
+    binding.readTestInitialLifecycleStateFromNativeWindow();
+    // Reset the state to detached. Going to paused first makes it a valid
+    // transition from any state.
+    await setAppLifeCycleState(AppLifecycleState.paused);
+    await setAppLifeCycleState(AppLifecycleState.detached);
+  });
+
+  testWidgets('listens to AppLifecycleState', (WidgetTester tester) async {
+    AppLifecycleState lastState = tester.binding.lifecycleState!;
+    void stateChange(AppLifecycleState state) {
+      lastState = state;
+    }
+
+    final AppLifecycleListener listener = AppLifecycleListener(
+      binding: WidgetsBinding.instance,
+      onStateChange: stateChange,
+    );
+    expect(lastState, equals(AppLifecycleState.detached));
+    await setAppLifeCycleState(AppLifecycleState.inactive);
+    expect(lastState, equals(AppLifecycleState.inactive));
+    await setAppLifeCycleState(AppLifecycleState.resumed);
+    expect(lastState, equals(AppLifecycleState.resumed));
+    listener.dispose();
+  });
+
+  testWidgets('Triggers correct state transition callbacks', (WidgetTester tester) async {
+    AppLifecycleState lastState = tester.binding.lifecycleState!;
```

Comment:
Should the test assert more on lastState (or should the onStateChange handler maybe be removed entirely since its covered by the previous test?). Currently it only asserts that its initial state is set correctly and doesn't assert that onStateChange is called.

Reply:
I just removed it.